include/picongpu/plugins/saxs/SaxsWriter.hpp

```diff
@@ -0,0 +1,496 @@
+        openPMD::Datatype datatype;
+        //! Output SI unit
+        const float_64 unit;
+        //! GridSpacing
+        float2_X const gridSpacing;
+
+
+public:
+
+
+        /** Initializes a SaxsWriter object.
+         *
+         * @param fileName Output file name, without the  extensions.
+         * @param fileExtension File extension, specifies the API backend.
+         * @param dir Where to save the output file.
+         * @param outputMemoryLayout  Functioning mode.
+         * @param compressionMethod
+         * @param globalExtent Output dimensions.
+         */
+        HINLINE SaxsWriter(
+            std::string const fileName,
+            std::string const fileExtension,
+            std::string const dir,
+            OutputMemoryLayout  outputMemoryLayout,
+            std::string const compressionMethod,
+            pmacc::math::UInt64< DIM2 > const globalExtent,
+            float2_X const gridSpacing,
+            float_64 const unit
+        ) :
+            fileName( fileName ),
+            dir( dir ),
+            fileExtension( fileExtension ),
+            outputMemoryLayout( outputMemoryLayout),
+            compressionMethod (compressionMethod),
+            globalExtent( globalExtent ),
+            gridSpacing(  gridSpacing  ),
+            unit( unit )
+        {
+            if(outputMemoryLayout ==
+                OutputMemoryLayout::SplitOverNodes)
+            {
+                // Set the MPI communicator.
+                GridController< simDim > & gc =
+                    Environment< simDim >::get().GridController();
+                __getTransactionEvent( ).waitForFinished( );
+                mpiCommunicator = MPI_COMM_NULL;
+                MPI_CHECK( MPI_Comm_dup(
+                    gc.getCommunicator( ).getMPIComm( ),
+                    & mpiCommunicator
+                ) );
+            }
+
+            datatype = openPMD::determineDatatype< T_ValueType >( );
+            // Create the output file.
+            openSeries( openPMD::AccessType::CREATE );
+            openPMDSeries->setMeshesPath( "scatteringData" );
+            closeSeries( );
+        }
+
+        virtual ~SaxsWriter( )
+        {
+            if(outputMemoryLayout == OutputMemoryLayout::SplitOverNodes)
+            {
+                if( mpiCommunicator != MPI_COMM_NULL )
+                {
+                    // avoid deadlock between not finished pmacc tasks and mpi
+                    // blocking collectives
+                    __getTransactionEvent( ).waitForFinished( );
+                    MPI_CHECK_NO_EXCEPT(
+                        MPI_Comm_free( &( mpiCommunicator ) ) );
+                }
+            }
+        }
+
+private:
+
+
+        inline bool isADIOS1()
+        {
+#if openPMD_HAVE_ADIOS1 && !openPMD_HAVE_ADIOS2
+            return this->fileExtension == "bp";
+#else
+            return false;
+#endif
+        }
+
+
+        //! Converts a pmacc Vector to an std::Vector without copying data.
```

Comment:
The method seems to actually copy data.

Reply:
That's true :D Well I took it from  #2966. But  maybe it won't copy data when the parameter is passed as an y-vaule  with std::move and that's how it was used in that PR hence the comment ?  But I'm using it only for members here so maybe I should just store it as an `std::vector` from the beginning. What do you think @sbastrakov ? :)

Reply:
I would just drop `without copying data.` from the comments. It is true that you can implement this logic so that is pushes back for for l-values and emplaces back for r-values, but here is seems really unnecessary, as this seems to be not a hotspot whatsoever, and also you use it just for floats, not any complicated type. So I would just edit the comment and leave the method simple. About storing as `std::vector` directly as member: you can do that as well, I do not have a particular preference over one way or another, think both are fine here.